scripts/tools/generate_esp32_chip_factory_bin.py

```diff
@@ -0,0 +1,139 @@
+        if ((args.passcode < 0x0000001 and args.passcode > 0x5F5E0FE) or (args.passcode in INVALID_PASSCODES)):
+            logging.error('Invalid passcode:' + str(args.passcode))
+            sys.exit(1)
+
+    # Validate the discriminator
+    if (args.discriminator is not None) and (args.discriminator not in range(0x0000, 0x0FFF)):
+        logging.error('Invalid discriminator:' + str(args.discriminator))
+        sys.exit(1)
+
+    logging.info('Discriminator:{} Passcode:{}'.format(args.discriminator, args.passcode))
+
+
+def gen_spake2p_params(passcode):
+    iter_count_min = 1000
+    iter_count_max = 10000
+    salt_len_min = 16
+    salt_len_max = 32
+
+    cmd = [
+        TOOLS['spake2p'], 'gen-verifier',
+        '--iteration-count', str(random.randint(iter_count_min, iter_count_max)),
+        '--salt-len', str(random.randint(salt_len_min, salt_len_max)),
+        '--pin-code', str(passcode),
+        '--out', '-',
+    ]
+
+    output = subprocess.check_output(cmd)
+    output = output.decode('utf-8').splitlines()
+    return dict(zip(output[0].split(','), output[1].split(',')))
+
+
+def generate_nvs_bin(discriminator, spake2p_params):
+    iteration = spake2p_params['Iteration Count']
+    salt = spake2p_params['Salt']
+    verifier = spake2p_params['Verifier']
+
+    csv_content = 'key,type,encoding,value\n'
+    csv_content += 'chip-factory,namespace,,\n'
+    csv_content += 'discriminator,data,u32,{}\n'.format(discriminator)
+    csv_content += 'iteration-count,data,u32,{}\n'.format(iteration)
+    csv_content += 'salt,data,string,{}\n'.format(salt)
+    csv_content += 'verifier,data,string,{}\n'.format(verifier)
+
+    with open('partition.csv', 'w') as f:
+        f.write(csv_content)
+
+    nvs_args = SimpleNamespace(input='partition.csv',
+                               output='partition.bin',
+                               size='0x6000',
+                               outdir=os.getcwd(),
+                               version=2)
+    nvs_partition_gen.generate(nvs_args)
+
+
+def print_flashing_help():
+    logging.info('To flash the generated partition.bin, run the following command:')
+    logging.info('==============================================================')
+    logging.info('esptool.py -p <port> write_flash <addr> partition.bin')
+    logging.info('==============================================================')
+    logging.info('default \"nvs\" partition addr is 0x9000')
+
+
+def main():
+    def any_base_int(s): return int(s, 0)
+
+    parser = argparse.ArgumentParser(description='Chip Factory NVS binary generator tool')
+    parser.add_argument('-p', '--passcode', type=any_base_int, default=20202021,
+                        help='The discriminator for pairing. Default is 20202021 if not specified.')
+    parser.add_argument('-d', '--discriminator', type=any_base_int, default=3840,
+                        help='The passcode for pairing. Default is 3840 if not specified.')
```

Comment:
Please do NOT provide defaults for these. It's a recipe to have those "SDK defaults" appear in the wild